src/box/memtx_engine.cc

```diff
@@ -333,7 +363,12 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	RegionGuard region_guard(&fiber()->gc);
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
-	*is_space_system = (request.space_id < BOX_SYSTEM_ID_MAX);
+	if (!space_id_is_system(request.space_id)) {
+		if (complete_system_space_recovery(state) != 0)
+			return -1;
```

Comment:
This return statement isn't covered by any tests. Please add a test for the case when a snapshot starts with a user space row.

Reply:
Added two more test cases. Actually, removing the check at the end of `memtx_engine_recovery_snapshot` was wrong — it misses the cases when the snapshot is completely empty.